### PR TITLE
Precompiled contracts for addition and scalar multiplication on the elliptic curve alt_bn128

### PR DESCRIPTION
**Pull request by chriseth**
Precompiled contracts for elliptic curve operations are required in order to perform zkSNARK verification within the block gas limit.

Replaces https://github.com/ethereum/EIPs/issues/196

### CODE REVIEW

File: EIPS/ecopts.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+The curve is defined by:
+```
+Y^2 = X^3 + 3
+over the field F_p with
+p = 21888242871839275222246405745257275088696311157297823662689037894645226208583
+```
+
+### Encoding
+
+Field elements are encoded as 32 byte big-endian numbers. Curve points are encoded as two field elements `(x, y)`, where the point at infinity is encoded as `(0, 0)`.
```
Comment (axic) [inline EIPS/ecopts.md]: Should also mention how is a scalar encoded.

File: EIPS/ecopts.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+For both precompiled contracts, if the input is shorter than expected, it is padded with zeros at the end.
+
+The length of the returned data is always as specified (i.e. it is not "unpadded"). 
+
+### Exact semantics
+
+Invalid input: For both contracts, if any input point does not lie on the curve or any of the field elements (point coordinates or scalar) is equal or larger than the field modulus p, the contract fails.
+
+ADD: Input: two curve points `(x, y)`. Fail on invalid input. Otherwise, return the curve point `x + y` where `+` is point addition on the elliptic curve `alt_bn128` specified above.
```
Comment (axic) [inline EIPS/ecopts.md]: Probably better to format it as:
```
#### ADD
Input: two curve points `(x, y)`...
Output: resulting curve point
...description...
```

File: EIPS/ecopts.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+## Motivation
+
+Current smart contract executions on Ethereum are fully transparent, which makes them unsuitable for several use-cases that involve private information like the location, identity or history of past transactions. The technology of zkSNARKs could be a solution to this problem. While the Ethereum Virtual Machine can make use of zkSNARKs in theory, they are currently too expensive
+to fit the block gas limit. Because of that, this EIP proposes to specify certain parameters for some elementary primitives that enable zkSNARKs so that they can be implemented more efficiently and the gas cost be reduced.
+
+Note that fixing these parameters will in no way limit the use-cases for zkSNARKs, it will even allow for incorporating some advances in zkSNARK research without the need for a further hard fork.
```
Comment (pirapira) [inline EIPS/ecopts.md]: I cannot immediately grasp how fixing parameters makes it easier to incorporate advances.

File: EIPS/ecopts.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+The curve is defined by:
+```
+Y^2 = X^3 + 3
+over the field F_p with
+p = 21888242871839275222246405745257275088696311157297823662689037894645226208583
+```
+
+### Encoding
+
+Field elements and scalars are encoded as 32 byte big-endian numbers. Curve points are encoded as two field elements `(x, y)`, where the point at infinity is encoded as `(0, 0)`.
```
Comment (pirapira) [inline EIPS/ecopts.md]: Isn't zero already used?  I guess the elements of F_p are represented by a number in the range of `[0, that prime number - 1]`.

File: EIPS/ecopts.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+For both precompiled contracts, if the input is shorter than expected, it is padded with zeros at the end.
```
Comment (pirapira) [inline EIPS/ecopts.md]: Is the padding bit-wise or byte-wise?

Comment (pirapira) [inline EIPS/ecopts.md]: Also, does it make sense to pad a big-endian number at the end?   `0x01` and `0x0100` would look the same.

File: EIPS/ecopts.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+The length of the returned data is always as specified (i.e. it is not "unpadded"). 
+
+### Exact semantics
+
+Invalid input: For both contracts, if any input point does not lie on the curve or any of the field elements (point coordinates or scalar) is equal or larger than the field modulus p, the contract fails.
+
+#### ADD
+Input: two curve points `(x, y)`.
```
Comment (pirapira) [inline EIPS/ecopts.md]: Does this mean the input has 128 bytes?

File: EIPS/ecopts.md (hunk 4 above)
Comment (chriseth) [inline EIPS/ecopts.md]: Yes, but `(0,0)` should not satisfy the curve equation.

Comment (pirapira) [inline EIPS/ecopts.md]: OK.  Somehow I was thinking about `(0, y)`s that might satisfy the equations.  Now I realize there `(0, y)`s are just interpreted as they are when `y` is not zero.

File: EIPS/ecopts.md (hunk 3 above)
Comment (chriseth) [inline EIPS/ecopts.md]: Changed the wording.

File: EIPS/ecopts.md (hunk 5 above)
Comment (chriseth) [inline EIPS/ecopts.md]: Changed the wording, please see if you come to the conclusion that these questions are irrelevant :-)

File: EIPS/ecopts.md (hunk 6 above)
Comment (chriseth) [inline EIPS/ecopts.md]: Clarified above.

File: EIPS/ecopts.md (hunk 5 above)
Comment (pirapira) [inline EIPS/ecopts.md]: I was reading "if any parameter is less than 2^255, thus any parameter is shorter than expected, the caller should pad the input at the end before calling the contract".  Now this interpretation is impossible.

File: EIPS/ecopts.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
+## Preamble
+<pre>
```
Comment (axic) [inline EIPS/ecopts.md]: Why the `<pre>`?

Comment (chriseth) [inline EIPS/ecopts.md]: That was part of the template.

Comment (axic) [inline EIPS/ecopts.md]: You might be using an old template.

**Comment by chriseth** [comment]
TODO: change semantics to throw an exception both for too short and too long input.

**Comment by chfast** [comment]
A comment from my perspective. @chriseth and me are in the process of "implementing" this and #212 EIPs. "Implementing" means to integrate the external library: [libff](https://github.com/scipr-lab/libff) which was a part of bigger [libsnark](https://github.com/scipr-lab/libsnark) just a month ago. The libff supports only Linux at the moment and nothing indicates that it is activelly developed. After the split the libff does not have tests properly integrated. It also depends on GMP what is another issue.

My personal opinion is this piece of code if to be integrated to any Ethereum client needs full code review and good set of unit tests.

I don't know why alt_bn128 curve is so special, but if we plan to stick with only this one for a long time, we can modify the code base of libff to be less generic and support only this single curve. Such fork can be joint effort of multiple teams if others are interested.

I wander what other libraries do you use. And what is used by ZCash?

**Comment by chriseth** [comment]
@chfast at least in the latest version, zcash uses libsnark (or some fork of it). Note that geth will use a different library, I think it might be this one here: https://github.com/golang/crypto/tree/master/bn256 and there is yet another implementation in rust: https://github.com/zcash/bn

So I think this is a C++-only discussion. We might take libff and swap out GMP for something else.

File: EIPS/ecopts.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+Current smart contract executions on Ethereum are fully transparent, which makes them unsuitable for several use-cases that involve private information like the location, identity or history of past transactions. The technology of zkSNARKs could be a solution to this problem. While the Ethereum Virtual Machine can make use of zkSNARKs in theory, they are currently too expensive
+to fit the block gas limit. Because of that, this EIP proposes to specify certain parameters for some elementary primitives that enable zkSNARKs so that they can be implemented more efficiently and the gas cost be reduced.
+
+Note that while fixing these parameters might look like limiting the use-cases for zkSNARKs, the primitives are so basic that they can be combined in ways that are flexible enough so that it should even be possible to allow future advances in zkSNARK research without the need for a further hard fork.
+
+## Specification
+
+Add precompiled contracts for point addition (ADD)  and scalar multiplication (MUL) on the elliptic curve "alt_bn128".
+
+Address of ADD: 0x6
```
Comment (pirapira) [inline EIPS/ecopts.md]: EIP 101 also uses this address.

File: EIPS/ecopts.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+Address for MUL: 0x7
```
Comment (pirapira) [inline EIPS/ecopts.md]: EIP 101 uses this address too.

Comment (chfast) [inline EIPS/ecopts.md]: Can we make sure the precompiled addresses are continuous in the end?

Comment (pirapira) [inline EIPS/ecopts.md]: I don't think so. `0x10` and `0x20` are taken in #96 .

Comment (pirapira) [inline EIPS/ecopts.md]: The available addresses seem to be `0xa` and `0xb`.

Comment (axic) [inline EIPS/ecopts.md]: Would it make sense changing #96 to use continuous addresses?

File: EIPS/ecopts.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+Inputs to test:
+
+ - Curve points which would be valid if the numbers were taken mod p (should fail).
+ - Both contracts should succeed on empty input.
+ - Truncated input that results in a valid curve point.
+ - Points not on curve (but valid otherwise).
+ - Multiply point with scalar that lies between the order of the group and the field (should succeed).
+ - Multiply point with scalar that is larger than the field order (should succeed).
+
```
Comment (axic) [inline EIPS/ecopts.md]: Can someone create those tests cases and include them here?

File: EIPS/ecopts.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+p = 21888242871839275222246405745257275088696311157297823662689037894645226208583
+```
+
+### Encoding
+
+Field elements and scalars are encoded as 32 byte big-endian numbers. Curve points are encoded as two field elements `(x, y)`, where the point at infinity is encoded as `(0, 0)`.
+
+Tuples of objects are encoded as their concatenation.
+
+For both precompiled contracts, if the input is shorter than expected, it is assumed to be virtually padded with zeros at the end (i.e. compatible with the semantics of the `CALLDATALOAD` opcode).
```
Comment (pirapira) [inline EIPS/ecopts.md]: What should happen when the input is longer than expected?

File: EIPS/ecopts.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+Fails on invalid input and consumes all gas provided.
+
+#### MUL
+Input: curve point and scalar `(x, s)`.
+Output: curve point `s * x`, where `*` is the scalar multiplication on the elliptic curve `alt_bn128` specified above.
+Fails on invalid input and consumes all gas.
+
+### Gas costs
+
+To be determined.
```
Comment (pirapira) [inline EIPS/ecopts.md]: Three or four clients have already implemented this.  So cross-benchmark should be possible already, for determining the ga s costs.

**Comment by pirapira** [comment]
@chriseth I'm waiting for the gas costs here.

**Comment by JustinDrake** [comment]
The Python link (https://github.com/ethereum/research/blob/master/zksnark/bn128_pairing.py) is broken.

**Comment by JustinDrake** [comment]
The C++ link https://github.com/scipr-lab/libsnark/blob/master/src/algebra/curves/alt_bn128/alt_bn128_g1.hpp is also broken.

**Comment by chriseth** [comment]
In my opinion, the only thing left to do here is to decide about gas costs.

**Comment by vbuterin** [comment]
@JustinDrake python link should point to https://github.com/ethereum/research/tree/master/zksnark/py_pairing/py_pairing

**Comment by chfast** [comment]
## Benchmarks
- Lower bound for average of random **ECMUL** is **~940000** CPU cycles (using libff + MPIR with native assembly).
- Lower bound for average of random **ECADD** is **~30000** CPU cycles (using libff + MPIR with native assembly).

**Comment by JustinDrake** [comment]
@vbuterin: That link is also broken https://github.com/ethereum/research/tree/master/zksnark/py_pairing/py_pairing

File: EIPS/ecopts.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+ - Points not on curve (but valid otherwise).
+ - Multiply point with scalar that lies between the order of the group and the field (should succeed).
+ - Multiply point with scalar that is larger than the field order (should succeed).
+
+## Implementation
+
+Implementation of these primitives are available here:
+
+ - [libff](https://github.com/scipr-lab/libff/blob/master/src/algebra/curves/alt_bn128/alt_bn128_g1.cpp) (C++)
+ - [bn](https://github.com/zcash/bn/blob/master/src/groups/mod.rs) (Rust)
```
Comment (gasolin) [inline EIPS/ecopts.md]: link is broken

File: EIPS/ecopts.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+In both codebases, a specific group on the curve alt_bn128 is used and is called G1.
+
+ - [Python](https://github.com/ethereum/research/blob/master/zksnark/py_pairing/py_pairing/bn128_curve.py) - probably most self-contained and best readable.
```
Comment (gasolin) [inline EIPS/ecopts.md]: link is broken

File: EIPS/ecopts.md (hunk 13 above)
Comment (pirapira) [inline EIPS/ecopts.md]: To be precise, the `libff` link is broken @chriseth

**Comment by axic** [comment]
I am 100% too late to the party with this, but it could be actually useful to have the inputs ABI encoded and reside on the same address for this reason: in WASM this precompile can be implemented natively as a contract, but with the current API it would be present in two copies.

File: EIPS/eip-213.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+## Test Cases
+
+Inputs to test:
+
+ - Curve points which would be valid if the numbers were taken mod p (should fail).
+ - Both contracts should succeed on empty input.
+ - Truncated input that results in a valid curve point.
+ - Points not on curve (but valid otherwise).
+ - Multiply point with scalar that lies between the order of the group and the field (should succeed).
+ - Multiply point with scalar that is larger than the field order (should succeed).
```
Comment (cdetrio) [inline EIPS/eip-213.md]: * YES
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/ecmul_7827-6598_9935_28000_96Filler.json#L137
    * multiplies by scalar (2**256 - 1) 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff

File: EIPS/eip-213.md [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+
+## Test Cases
+
+Inputs to test:
+
+ - Curve points which would be valid if the numbers were taken mod p (should fail).
+ - Both contracts should succeed on empty input.
+ - Truncated input that results in a valid curve point.
+ - Points not on curve (but valid otherwise).
+ - Multiply point with scalar that lies between the order of the group and the field (should succeed).
```
Comment (cdetrio) [inline EIPS/eip-213.md]: * YES
  * order of the group: curve_order = 21888242871839275222246405745257275088548364400416034343698204186575808495617
    - in hex: 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
  * order of the field (same as modulus of the field):  field_modulus = 21888242871839275222246405745257275088696311157297823662689037894645226208583
    - in hex: 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47

  https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/pointMulAdd2Filler.json#L692-L693
  * uses ECMUL with scalars -1 mod field_order and -2 mod field_order (0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd46 and 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd45)

File: EIPS/eip-213.md [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+As with the introduction of any precompiled contract, contracts that already use the given addresses will change their semantics. Because of that, the addresses are taken from the "reserved range" below 256.
+
+## Test Cases
+
+Inputs to test:
+
+ - Curve points which would be valid if the numbers were taken mod p (should fail).
+ - Both contracts should succeed on empty input.
+ - Truncated input that results in a valid curve point.
+ - Points not on curve (but valid otherwise).
```
Comment (cdetrio) [inline EIPS/eip-213.md]: * YES
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/ecadd_0-3_1-2_25000_128Filler.json
    * (0,3) is not on curve
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/ecmul_1-3_1_28000_128Filler.json#L137
    * (1,3) is not on the curve

File: EIPS/eip-213.md [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+
+As with the introduction of any precompiled contract, contracts that already use the given addresses will change their semantics. Because of that, the addresses are taken from the "reserved range" below 256.
+
+## Test Cases
+
+Inputs to test:
+
+ - Curve points which would be valid if the numbers were taken mod p (should fail).
+ - Both contracts should succeed on empty input.
+ - Truncated input that results in a valid curve point.
```
Comment (cdetrio) [inline EIPS/eip-213.md]: * NO
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/pointAddTruncFiller.json
    - calls ECADD but only truncates at the 64th byte, i.e. after a full point. does not truncate at e.g. 96th byte (after the x coord of the 2nd point) or at bytes in the middle of a coordinate
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/ecadd_1-3_0-0_25000_80Filler.json
    - calls ECADD but truncates at the 80th byte of a zero point
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/ecmul_7827-6598_1456_28000_80Filler.json
    - truncates at the 80th byte, which is the middle of the scalar factor, not the middle of a point

File: EIPS/eip-213.md [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+## Backwards Compatibility
+
+As with the introduction of any precompiled contract, contracts that already use the given addresses will change their semantics. Because of that, the addresses are taken from the "reserved range" below 256.
+
+## Test Cases
+
+Inputs to test:
+
+ - Curve points which would be valid if the numbers were taken mod p (should fail).
+ - Both contracts should succeed on empty input.
```
Comment (cdetrio) [inline EIPS/eip-213.md]: * YES
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/ecadd_0-0_0-0_21000_0Filler.json#L135
    * calls ECADD with empty input. 
    * the tx data 0x30c8d1da00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000 has 0x20 to encode the len(size), and 0x00..00 as the size of the input.
  * https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stZeroKnowledge/ecmul_0-0_0_28000_0Filler.json
    * calls ECMUL with empty input

File: EIPS/eip-213.md (hunk 18 above)
Comment (pirapira) [reply EIPS/eip-213.md]: Added such cases in the test [google spreadsheet](https://docs.google.com/spreadsheets/d/1xat7UI8GtB4ZGVdlK5_XQSHJZaMThi4SrlcL8XMZb5Q/edit#gid=941134795&range=B13:B15)

File: EIPS/eip-213.md (hunk 15 above)
Comment (pirapira) [reply EIPS/eip-213.md]: @cdetrio do you think a link to the test filler should be added to the EIP?

Comment (cdetrio) [reply EIPS/eip-213.md]: @pirapira I was intending to link the test cases. But on second thought, we don't do that for other EIPs so I guess having the links in the PR/discussion is enough.

**Comment by shamatar** [comment]
Hello @pirapira 

As important optimizations were merged in 1.8.x and actual processor time for every operation (especially MUL) was significantly reduced, what are the plans to reduce a precompile costs?

Sincerely, Alex

**Comment by pirapira** [comment]
@shamatar thanks for the news.  Many people would be happier with the lower gas amounts, so my suggestion is to file an EIP about the newer gas costs (if there is none yet).

**Comment by Shadowfiend** [comment]
I've opened #1088 to propose this, as I didn't see any other EIPs on the matter.

**Pull request merged**